"Filesystem changed during run" doesn't restart tests that have completed but changed

**Describe the bug**

Pants' automatic detection of changes and thus ability to rerun things as required is cool, however it seems to be very coarse grained. For instance, if running `./pants test ::`, and a test finishes, then it won't be rerun by that invocation of pants, even if the file system changes in way that would affect it. Instead, it seems to only be tasks that are still running that are restarted (even if the change doesn't affect them).

https://gist.github.com/huonw/40492d86e8ac922327b1106dc13b4a8a has a demo: the `run-it.sh` script does the following:

1. creates two test files `test_slow.py` (takes 10s) and `test_fast.py` (instantaneous)
2. starts `./pants test ::` (would take 10s due to `test_slow.py`)
3. after 5s (i.e. only `test_slow.py` is running) rewrites `test_fast.py`, in a way that will fail.
4. BUG: the rewrite causes `test_slow.py` to be interrupted and rerun, while `test_fast.py` doesn't rerun. That test run completes (after about 15s)
5. runs `./pants test ::` again, which uses the memoised result from `test_slow.py` (yay), but reruns `test_fast.py` and only now observes the failure

```shell
git clone git@gist.github.com:40492d86e8ac922327b1106dc13b4a8a.git
cd 40492d86e8ac922327b1106dc13b4a8a
./run-it.sh
```
Trimmed output (timestamp `09:26:45.07` is the bug):

```
+ write_file test_fast.py 0
...
+ write_file test_slow.py 10
...

+ ./pants version
09:26:39.86 [INFO] Initializing scheduler...
09:26:39.98 [INFO] Scheduler initialized.
2.14.0

+ ./pants test ::
09:26:41.97 [INFO] Completed: Run Pytest - //test_fast.py:test succeeded.
09:26:45.07 [INFO] Filesystem changed during run: retrying `Test` in 500ms...
09:26:55.82 [INFO] Completed: Run Pytest - //test_slow.py:test succeeded.

✓ //test_fast.py:test succeeded in 0.18s.
✓ //test_slow.py:test succeeded in 10.23s.

+ ./pants test ::
09:26:56.51 [INFO] Completed: Run Pytest - //test_slow.py:test succeeded.
09:26:56.82 [ERROR] Completed: Run Pytest - //test_fast.py:test failed (exit code 2).
============================= test session starts ==============================
...
=============================== 1 error in 0.11s ===============================


✓ //test_slow.py:test succeeded in 10.23s (memoized).
✕ //test_fast.py:test failed in 0.28s.
```


I'd be expecting step 4 to leave `test_slow.py` running, and, most critically, rerun `test_fast.py` straight away, meaning the failure is detected in the first `./pants test ::` run.

**Pants version**
2.14.0

**OS**
macOS

**Additional info**
https://gist.github.com/huonw/40492d86e8ac922327b1106dc13b4a8a

--- a/src/rust/engine/graph/src/entry.rs
+++ b/src/rust/engine/graph/src/entry.rs
@@ -122,10 +122,12 @@ impl<N: Node> EntryResult<N> {
   /// If the value is in a Clean state, mark it Dirty.
   fn dirty(&mut self) {
     match self {
-      EntryResult::Clean(v) | EntryResult::UncacheableDependencies(v, _) => {
+      EntryResult::Clean(v)
+      | EntryResult::UncacheableDependencies(v, _)
+      | EntryResult::Uncacheable(v, _) => {
         *self = EntryResult::Dirty(v.clone());
       }
-      EntryResult::Dirty(_) | EntryResult::Uncacheable(_, _) => {}
+      EntryResult::Dirty(_) => {}
     }
   }
 

--- a/src/rust/engine/graph/src/lib.rs
+++ b/src/rust/engine/graph/src/lib.rs
@@ -280,6 +280,12 @@ impl<N: Node> InnerGraph<N> {
       dirtied: transitive_ids.len(),
     };
 
+    // If there were no roots, then nothing will be invalidated. Return early to avoid scanning all
+    // edges in `retain_edges`.
+    if root_ids.is_empty() {
+      return invalidation_result;
+    }
+
     // Clear roots and remove their outbound edges.
     for id in &root_ids {
       if let Some(entry) = self.pg.node_weight_mut(*id) {

--- a/src/rust/engine/graph/src/node.rs
+++ b/src/rust/engine/graph/src/node.rs
@@ -43,13 +43,18 @@ pub trait Node: Clone + Debug + Display + Eq + Hash + Send + 'static {
   /// If a node's output is cacheable based solely on properties of the node, and not the output,
   /// return true.
   ///
+  /// Nodes which are not cacheable will be recomputed once (at least, in case of dirtying) per
+  /// RunId.
+  ///
   /// This property must remain stable for the entire lifetime of a particular Node, but a Node
   /// may change its cacheability for a particular output value using `cacheable_item`.
   ///
   fn cacheable(&self) -> bool;
 
+  ///
   /// A Node may want to compute cacheability differently based on properties of the Node's item.
   /// The output of this method will be and'd with `cacheable` to compute overall cacheability.
+  ///
   fn cacheable_item(&self, _item: &Self::Item) -> bool {
     self.cacheable()
   }

--- a/src/rust/engine/graph/src/tests.rs
+++ b/src/rust/engine/graph/src/tests.rs
@@ -100,6 +100,53 @@ async fn invalidate_and_rerun() {
 }
 
 #[tokio::test]
+async fn invalidate_uncacheable() {
+  let graph = empty_graph();
+
+  // Create three nodes, with the middle node as uncacheable.
+  let context = {
+    let mut uncacheable = HashSet::new();
+    uncacheable.insert(TNode::new(1));
+    TContext::new(graph.clone()).with_uncacheable(uncacheable)
+  };
+  assert_eq!(
+    graph.create(TNode::new(2), &context).await,
+    Ok(vec![T(0, 0), T(1, 0), T(2, 0)])
+  );
+  assert_eq!(
+    context.runs(),
+    vec![TNode::new(2), TNode::new(1), TNode::new(0)]
+  );
+
+  // Clear the bottom Node, which dirties the middle and upper node.
+  assert_eq!(
+    graph.invalidate_from_roots(true, |n| n.id == 0),
+    InvalidationResult {
+      cleared: 1,
+      dirtied: 2
+    }
+  );
+
+  // Re-request in the same session with new salt, and validate that all three re-run.
+  let context = context.with_salt(1);
+  assert_eq!(
+    graph.create(TNode::new(2), &context).await,
+    Ok(vec![T(0, 1), T(1, 1), T(2, 1)])
+  );
+  assert_eq!(
+    context.runs(),
+    vec![
+      TNode::new(2),
+      TNode::new(1),
+      TNode::new(0),
+      TNode::new(1),
+      TNode::new(0),
+      TNode::new(2)
+    ]
+  );
+}
+
+#[tokio::test]
 async fn invalidate_with_changed_dependencies() {
   let graph = empty_graph();
   let context = TContext::new(graph.clone());

